File: backend/app/schemas/organizer.py
from pydantic import BaseModel, UUID4
from typing import List, Optional, Dict, Any
from datetime import datetime

# --- Ingestion Schemas ---

class IngestRequest(BaseModel):
    text: str
    confirmed_category: str
    confirmed_parent: Optional[str] = None
    confirmed_type: Optional[str] = None
    confirmed_nature: Optional[str] = None
    source: str = "user"

# --- Feedback Schemas ---

class FeedbackRequest(BaseModel):
    text: str
    chosen_category: str
    chosen_parent: Optional[str] = None
    chosen_type: Optional[str] = None
    chosen_nature: Optional[str] = None

# --- Memory Schemas ---

class OrganizerMemoryBase(BaseModel):
    text: str
    normalized_text: str
    chosen_category: str
    chosen_parent: Optional[str] = None
    chosen_type: Optional[str] = None
    chosen_nature: Optional[str] = None
    source: str

class OrganizerMemoryCreate(OrganizerMemoryBase):
    pass

class OrganizerMemorySchema(OrganizerMemoryBase):
    id: UUID4
    created_at: datetime

    class Config:
        from_attributes = True

# --- Rule Schemas ---

class OrganizerRuleBase(BaseModel):
    rule_pattern: str
    suggested_category: str
    suggested_parent: Optional[str] = None
    confidence: float

class OrganizerRuleCreate(OrganizerRuleBase):
    pass

class OrganizerRuleSchema(OrganizerRuleBase):
    id: UUID4
    created_at: datetime

    class Config:
        from_attributes = True

# --- Classifier Schemas ---

class ClassificationResponseExtended(BaseModel):
    suggested_category: Optional[str] = None
    suggested_parent: Optional[str] = None
    suggested_type: Optional[str] = None
    suggested_nature: Optional[str] = None
    confidence: float
    similar_existing_accounts: List[Dict[str, Any]]
    raw_model_output: Optional[str] = None
    rule_engine_flags: Optional[List[str]] = None
    error: Optional[str] = None
    memory_votes: List[Dict[str, Any]] = []
    dynamic_rules_triggered: List[Dict[str, Any]] = []
    reasoning: Optional[str] = None
    final_weights: Dict[str, float] = {}

File: frontend/src/types/organizer.ts
// frontend/src/types/organizer.ts

export interface OrganizerClassification {
    suggested_category: string | null;
    suggested_parent: string | null;
    suggested_type: string | null;
    suggested_nature: string | null;
    confidence: number;
    similar_existing_accounts: any[]; // Define more strictly if needed
    raw_model_output: string | null;
    rule_engine_flags: string[] | null;
    error: string | null;
    memory_votes: any[];
    dynamic_rules_triggered: any[];
    model_reasoning: string | null;
    final_weights: Record<string, number>;
    suggested_code?: string;
}

export interface OrganizerMemory {
    id: string;
    text: string;
    normalized_text: string;
    chosen_category: string;
    chosen_parent: string | null;
    chosen_type: string | null;
    chosen_nature: string | null;
    source: string;
    created_at: string;
}

export interface OrganizerRule {
    id: string;
    rule_pattern: string;
    suggested_category: string;
    suggested_parent: string | null;
    confidence: number;
    created_at: string;
}
